consensus/istanbul/backend/internal/enodes/val_enode_db.go

```diff
@@ -212,78 +213,103 @@ func (vet *ValidatorEnodeDB) String() string {
+	vet.lock.RLock()
+	defer vet.lock.RUnlock()
+	var entries = make(map[common.Address]*AddressEntry)
+
+	err := vet.iterateOverAddressEntries(func(address common.Address, entry *AddressEntry) error {
+		entries[address] = entry
+		return nil
+	})
+
+	if err != nil {
+		vet.logger.Error("ValidatorEnodeDB.GetAllAddressEntries error", "err", err)
+		return nil, err
+	}
+
+	return entries, nil
+}
+
 // Upsert will update or insert a validator enode entry; given that the existing entry
 // is older (determined by view parameter) that the new one
-func (vet *ValidatorEnodeDB) Upsert(remoteAddress common.Address, enodeURL string, view *istanbul.View) error {
+func (vet *ValidatorEnodeDB) Upsert(valEnodeEntries map[common.Address]*AddressEntry) error {
 	vet.lock.Lock()
 	defer vet.lock.Unlock()
 
-	currentEntry, err := vet.getAddressEntry(remoteAddress)
-	isNew := err == leveldb.ErrNotFound
+	batch := new(leveldb.Batch)
+	peersToRemove := make([]*enode.Node, 0, len(valEnodeEntries))
+	peersToAdd := make(map[common.Address]*enode.Node)
 
-	// Check errors
-	if !isNew && err != nil {
-		return err
-	}
+	for remoteAddress, addressEntry := range valEnodeEntries {
+		currentEntry, err := vet.GetAddressEntry(remoteAddress)
+		isNew := err == leveldb.ErrNotFound
 
-	// If it is an old message, ignore it.
-	if err == nil && view.Cmp(currentEntry.view) <= 0 {
-		return errOldAnnounceMessage
-	}
+		// Check errors
+		if !isNew && err != nil {
+			return err
+		}
 
-	// new entry
-	rawEntry, err := rlp.EncodeToBytes(&addressEntry{enodeURL, view})
-	if err != nil {
-		return err
-	}
+		// new entry
+		rawEntry, err := rlp.EncodeToBytes(addressEntry)
+		if err != nil {
+			return err
+		}
 
-	hasOldValueChanged := !isNew && currentEntry.enodeURL == enodeURL
+		hasOldValueChanged := !isNew && currentEntry.Node.String() == addressEntry.Node.String()
 
-	batch := new(leveldb.Batch)
+		if hasOldValueChanged {
+			batch.Delete(nodeIDKey(currentEntry.Node.ID()))
+			batch.Put(nodeIDKey(addressEntry.Node.ID()), remoteAddress.Bytes())
+			peersToRemove = append(peersToRemove, currentEntry.Node)
+		} else if isNew {
+			batch.Put(nodeIDKey(addressEntry.Node.ID()), remoteAddress.Bytes())
+			peersToAdd[remoteAddress] = addressEntry.Node
+		}
+		batch.Put(addressKey(remoteAddress), rawEntry)
 
-	if hasOldValueChanged {
-		batch.Delete(enodeURLKey(currentEntry.enodeURL))
-		batch.Put(enodeURLKey(enodeURL), remoteAddress.Bytes())
-	} else if isNew {
-		batch.Put(enodeURLKey(enodeURL), remoteAddress.Bytes())
+		vet.logger.Trace("Going to upsert an entry in the valEnodeTable", "address", remoteAddress, "enodeURL", addressEntry.Node.String())
 	}
-	batch.Put(addressKey(remoteAddress), rawEntry)
 
-	err = vet.db.Write(batch, nil)
-	if err != nil {
-		return err
-	}
-	vet.logger.Trace("Upsert an entry in the valEnodeTable", "address", remoteAddress, "enodeURL", enodeURL)
+	if batch.Len() > 0 {
+		if err := vet.db.Write(batch, nil); err != nil {
```

Comment:
maybe it's idiomatic, but feels strange. the `else` of write continues with the operation...
what about?
```go
err := write...
if err { exit }
removeAndAddPeers...
```